

## README


This repository contains scripts and data to reproduce results of the article:

```
Pascual-Garc√≠a, A., Rivett, D., Jones, M.J., Bell, T., Replaying the tape of ecology to domesticate wild microbiota (2023)
```

### Organization of the repository

#### Folders
The repository has a folder containing all the scripts (`src`) and several folders for the data. It is necessary to keep this structure to execute the scripts. Folders labelled 1 to 6 are used sequentially for dada2 pipeline, with final ASV tables available in folder number 6. Folders labelled 7 contain analysis related to microbial composition. Folder 8 contains PICRUSt analysis and folder 9 the predictions obtained after optimal superposition.

#### Files excluded

To keep the repository as light as possible, a number of files were excluded. These files can be regenerated using the scripts provided.

* Folders 1 to 3 are completely empty because they should contain the raw sequencing data and a number of large output files from dada2. 
* The following files are also excluded (included in `.gitignore`):
    * `dada_output.RDS`, `seqtab.csv` and `seqtab.nochim.csv`.
    * `silva*`.
    * `4_dada2/DECIPHER_2.20.0`
    * `5_treeholes/seqtab_treeholes.csv`
    * `8_PICRUSt2/KO_metagenome_out/pred_metagenome_unstrat.tsv` (A bzip2 compressed file with the same name is available)
    * Intermediate files generated by PICRUSt (`8_PiCRUST/intermediate`).
    * All figures (extension `*pdf`).

### Input data

* Raw sequences are available in XXXX under accession number XXXX. Users interested in reproducing the whole pipeline should download the file(s) XXX in the folder(s) XXXX and then execute the dada2 pipeline (see scripts description below).
* Users interested in the processed data will find: 
    * `seqtable_readyforanalysis.csv`: ASV vs samples table, found in directory `6_finalfiles`.
    * `metadata_Time0D-7D-4M_May2022.csv`: starting metadata table, found in directory `4_dada2`. During the analysis, this table will be extended. Therefore, other metadata tables can be found in subsequent folders.
    * `taxa_wsp_readyforanalysis.csv`: Taxonomy, also found in directory `6_finalfiles`.

### Scripts

Most scripts have a header describing their usage. All of them were coded considering the structure of the repository, so there is no need to modify the paths, hence are hard-coded relative to the root of the repository. Please note that to make the scripts portable some functions to recover the user's path are used which, in the case of R scripts, require executing the script from Rstudio. Some scripts have different options for the analysis, indicated in variables. The main scripts are:

**Include here dada2 stuff**

* `main_find_classes.R` Computes the all-against-all JSD for all samples and looks for the optimal partition (output in `7.1_classes`).
* `main_find_classes_exp-split.R` Same analysis for each experiment and replicate independently (output in `7.1_classes`).
* `merge_metadata.R` Script to merge metadata tables obtained with the previous scripts (output in `7.1_classes`).
* `compare_classes.R` Generates a heatmap of the mean beta-diversity of each class and cluster classes (output in `7.1_classes`).
* `TraceSamples_fromPartTime0toPartTime7.pl` Performs statistics between starting and final classes (output in `7.2_match`).
* `match_clusters_time0to7_V2.R` Represent the above statistics (output in `7.2_match`).
* `phyloseq_analysis.R` Creates bar plots and some ordinations (output in `7.3_phyloseq`).
* `anosimlike_analysis.R` Provides a quantitative estimation of samples' similarity between the different categorical groups through the ANOSIM statistics (output in `7.4_variance`)
* `distance_to_attractor.R` Provides a quantitative estimation for the probability of having all replicates in the same class as a function of the distance to the nearest attractor of the final classes (output in `7.5_attractors`).
* `represent_attractors.R` Some visualizations of the results obtained with `distance_to_attractor.R` (output in `7.5_attractors`).
* `pathways_to_heatmap.R` Creates some plots of PICRUSt results (output in `8_PICRUSt`).
* `community_superposition.R` Looks for the optimal superposition between starting communities and one of the replicates, performs a prediction and compares with the remaining replicates (output in `9_predictions`).



